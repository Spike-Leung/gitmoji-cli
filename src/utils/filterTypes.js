// @flow
import Fuse from 'fuse.js'

import { type Types } from "@constants/types"

export const options = {
  threshold: 0.5,
  keys: [
    {
      name: 'name',
      weight: 0.5
    },
    {
      name: 'description',
      weight: 0.5
    }
  ]
}

const filterTypes = (
  input: ?string,
  types: Array<Types>
): Function | Array<Object> => {
  const fuse = new Fuse(types, options)

  return input ? fuse.search(input).map((type) => type.item) : types
}

export default filterTypes
