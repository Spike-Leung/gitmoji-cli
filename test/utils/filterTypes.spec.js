import filterTypes, { options } from '../../src/utils/filterTypes.js'
import TYPES from '../../src/constants/types.js'

describe('filterTypes', () => {
  it('should find all types with empty input', () => {
    const filteredTypes = filterTypes(undefined, TYPES)

    expect(filteredTypes).toStrictEqual(TYPES)
  })

  it('should should find the `feat` type first', () => {
    const filteredTypes = filterTypes('feat', TYPES)

    const type = TYPES.find((t) => t.name === 'feat')

    expect(filteredTypes[0]).toStrictEqual(type)
  })

  it('should should find the `test` type first', () => {
    const filteredTypes = filterTypes('test', TYPES)

    const type = TYPES.find((t) => t.name === 'test')

    expect(filteredTypes[0]).toStrictEqual(type)
  })

  it('should match a list of filteredTypes', () => {
    const filteredTypes = filterTypes('feat', TYPES)

    expect(filteredTypes).toMatchSnapshot()
  })

  it('should not have a total weight of greater than 1', () => {
    const weightSum = options.keys.reduce(
      (carry, key) => (carry += key.weight),
      0
    )

    expect(weightSum).toBeLessThanOrEqual(1)
  })
})
