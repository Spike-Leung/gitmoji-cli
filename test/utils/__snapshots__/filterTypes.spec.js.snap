// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterTypes should match a list of filteredTypes 1`] = `
[
  {
    "description": "A new feature",
    "name": "feat",
  },
  {
    "description": "Adding missing tests or correcting existing tests",
    "name": "test",
  },
  {
    "description": "Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)",
    "name": "build",
  },
  {
    "description": "A code change that neither fixes a bug nor adds a feature",
    "name": "refactor",
  },
  {
    "description": "Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)",
    "name": "style",
  },
]
`;
